qutebrowser/completion/completionwidget.py

```diff
@@ -263,6 +304,8 @@ def completion_item_focus(self, which, history=False):
             'prev': self._next_idx(upwards=True),
```

Comment:
You were just following the pattern here, but I think these should all be changed to lambdas. Right now it is calculating all of the indices and only using one of them. That is why the [test is failing](https://travis-ci.org/qutebrowser/qutebrowser/jobs/438241875#L702) when you haven't even added tests for 'next/last-page' yet.

IMHO an `if/elif` block would be fine too.

That being said, we still need to figure out why `self.visualRect(idx).height()` is sometimes `0`.

Reply:
Maybe you just need to set a font or something for the tests? I can't get it to fail when actually running the browser.

Reply:
Sorry, but what lines do you want implemented as lambdas?

Reply:
```py
    indices = {
        'next': lambda: self._next_idx(upwards=False),
        'prev': lambda: self._next_idx(upwards=True),
        'next-category': lambda: self._next_category_idx(upwards=False),
        'prev-category': lambda: self._next_category_idx(upwards=True),
        'next-page': lambda: self._next_page(upwards=False),
        'last-page': lambda: self._next_page(upwards=True),
    }
    idx = indices[which]()
```

or

```py
if which == 'next':
    idx = self._next_idx(upwards=False)
elif which == 'prev':
    idx = self._next_idx(upwards=True)
elif which == 'next-category':
    idx = self._next_category_idx(upwards=False)
elif which == 'prev-category':
    idx = self._next_category_idx(upwards=True)
elif which == 'next-page':
    idx = self._next_page(upwards=False)
elif which == 'last-page':
    idx = self._next_page(upwards=True)
```

So we aren't computing every index when we only want one.